### PR TITLE
Create a libflexpret

### PR DESCRIPTION
**Pull request by erlingrj**
1. Remove all function definitions from header files. This means adding `flexpret_io.c`. 
2. Add `flexpret.h` which is a single file to include to get all flexpret library functions in scope
3. Compile libflexpret as a static library and link it with the application. Currently I have added tinyalloc into libflexpret. 

I wonder if there are any reasons for not just ripping of the bandaid and moving to CMake. Since LF is CMake-based it will ease the integration there. Any thoughts on that?

### CODE REVIEW

**Comment by sberkun** [comment]
I started with a Makefile system because I thought it would be simpler. There are already existing Makefile-based things in this repo (i.e. the [emulator Makefile](https://github.com/pretis/flexpret/blob/master/Makefile) and the [tests Makefile](https://github.com/pretis/flexpret/blob/master/programs/tests/Makefile)), so I settled on a Makefile rather than trying to introduce CMake.

 In terms of integrating with LF CMake, I'm unsure how we would do so cleanly.  It looks like a flexpret+LF workflow would look like
  - lfc compiles `example.lf` into C files, invokes CMake to create the `.riscv` file, `.dump` file, and `.mem` file, then copies the `.mem` file into `bin/example.mem`
  - user invokes `fp-emu +ispm=example.mem` (which generates a `.vcd` in the cwd), or uploads the `.mem` to the fpga.

What I don't like about this is that it leaves the generated artifacts (`.riscv`, `.dump`, `.mem`, `.vcd`) all over the place, and the second step has to be invoked manually. 

I was thinking of a makefile-based workflow, which would look like:
 - lfc compiles `example.lf` into C files
 - user `cd`s into the generated source, then `make run` to compile and run on the emulator, or `make upload` to upload to the fpga.

File: programs/lib/include/flexpret_io.h [hunk 1]
```diff
@@ -1,58 +1,50 @@
...
-#include <flexpret_csrs.h>
-
 #ifndef FLEXPRET_IO_H
 #define FLEXPRET_IO_H
 
+#include <stdint.h>
+#include <flexpret_csrs.h>
+
 // Write a generic value to the tohost CSR
-static inline void write_tohost(uint32_t val) { write_csr(CSR_TOHOST, val); }
```
[the remark this answers is not in the document]
Comment (erlingrj) [reply programs/lib/include/flexpret_io.h:8]: Makes sense, will change back

**Comment by erlingrj** [comment]
There has been some refactorings to `reactor-c` and it is using CMake to build several(?) libraries which is linked with the auto-generated sources. Either we have to redo this using Make, or we have to use CMake. Making it work with CMake should be easy, it is just a matter of  changing the variables relating to the C compiler and C flags. You are right that the last steps requires some customization, but hopefully possible in CMake. 

Maybe Peter could comment on this.

**Comment by edwardcwang** [comment]
I would also prefer to support the usecase of using FlexPRET without LF as well...

**Comment by erlingrj** [comment]
Obviously we will support bare-metal C programming for FlexPRET, which is well supported by both Make and CMake. Anyways, I conclude that we keep the Makefile-based system because it works. It can also be a nice use-case to see how easy it is to integrate LF with non-CMake build systems.

**Comment by erlingrj** [comment]
I have made more changes now. I removed the old Makefile-based test-framework. It duplicated all the build logic that Samuel also have implemented. I think it is simplified considerably now.

File: programs/lib/include/flexpret_io.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (sberkun) [reply programs/lib/include/flexpret_io.h:8]: sorry, was this changed back?

**Comment by erlingrj** [comment]
I need a approving review to merge this  @lsk567 @sberkun

**Review by sberkun — APPROVED** [review]
lgtm!

**Pull request merged**